### PR TITLE
[MCH-251-117] Feat: crvYfiEth smart solver

### PR DESCRIPTION
**Pull request by gotzenx**
Now solver can execute even if only one token is ready to be swapped, and use the correct `execute` method on TF contract for gas optimization.

### CODE REVIEW

File: packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts [hunk 1]
```diff
@@ -52,81 +64,115 @@ export class CurveYfiEth implements Solver {
...
-    if (approveCrv) {
-      console.log('[CurveYfiEth] Approving crv');
-      const approveCrvTx = await this._crv.populateTransaction.approve(zrxCrvAllowanceTarget, constants.MaxUint256);
-      await multicallSwapperSigner.sendTransaction(approveCrvTx);
-      transactions.push(approveCrvTx);
+    // Create list of trades details to be executed
+    const asyncTradesExecutionDetails: AsyncTradeExecutionDetailsStruct[] = [];
+
+    console.log('[CurveYfiEth] CRV management');
+    const shouldExecuteCrvTrade = (await this._crv.balanceOf(strategy)).gt(DUST_THRESHOLD);
```
Comment (b0dhidharma) [inline packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts:81]:
```suggestion
    const shouldExecuteCrvTrade = crvBalance.gt(DUST_THRESHOLD);
```

File: packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts [hunk 2]
```diff
@@ -52,81 +64,115 @@ export class CurveYfiEth implements Solver {
...
-    });
-
-    const approveCvx = (await this._cvx.allowance(multicallSwapperAddress, zrxCvxAllowanceTarget)).lt(cvxBalance);
-    if (approveCvx) {
-      console.log('[CurveYfiEth] Approving cvx');
-      const approveCvxTx = await this._cvx.populateTransaction.approve(zrxCvxAllowanceTarget, constants.MaxUint256);
-      await multicallSwapperSigner.sendTransaction(approveCvxTx);
-      transactions.push(approveCvxTx);
+    console.log('[CurveYfiEth] CVX management');
+    const shouldExecuteCvxTrade = (await this._cvx.balanceOf(strategy)).gt(DUST_THRESHOLD);
```
Comment (b0dhidharma) [inline packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts:125]:
```suggestion
    const shouldExecuteCvxTrade = cvxBalance.gt(DUST_THRESHOLD);
```

**Review by b0dhidharma — CHANGES_REQUESTED** [review]
Other than the suggested changes this is 10/10. Great f* work !

File: packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts [hunk 3]
```diff
@@ -52,81 +64,115 @@ export class CurveYfiEth implements Solver {
...
-    });
-
-    const approveCvx = (await this._cvx.allowance(multicallSwapperAddress, zrxCvxAllowanceTarget)).lt(cvxBalance);
-    if (approveCvx) {
-      console.log('[CurveYfiEth] Approving cvx');
-      const approveCvxTx = await this._cvx.populateTransaction.approve(zrxCvxAllowanceTarget, constants.MaxUint256);
-      await multicallSwapperSigner.sendTransaction(approveCvxTx);
-      transactions.push(approveCvxTx);
+    console.log('[CurveYfiEth] CVX management');
+    const shouldExecuteCvxTrade = cvxBalance.gt(DUST_THRESHOLD);
```
Comment (poolpitako) [inline packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts:125]: I don't know if it makes sense to execute these trades separately.
Perhaps something like:
```
cvxBalance.gt(DUST_THRESHOLD) || crvBalance.gt(DUST_THRESHOLD)
```

makes more sense

**Review by poolpitako — APPROVED** [review]
This refactors looks good. At a certain point we should abstract some of the logic and apply to the other trade which is almost the same: spell/eth.

File: packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts (hunk 3 above)
Comment (poolpitako) [reply packages/yswaps/scripts/libraries/solvers/multicall/CurveYfiEth.ts:125]: they should increase together since it's a convex strat.

**Pull request merged**